In [8]:
import pandas as pd
df = pd.read_csv('heart.csv')

In [9]:
X = df.drop('target', axis=1)
y = df['target']

In [10]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [11]:
df.shape

(1025, 14)

In [12]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

In [13]:
model = LogisticRegression()

In [14]:
scores = cross_val_score(model, X_scaled, y, cv=5)

In [15]:
print(scores)

[0.88292683 0.85365854 0.86341463 0.82439024 0.80487805]


In [16]:
print(scores.mean())

0.8458536585365855


In [17]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

In [18]:
lr = LogisticRegression()
lr_scores = cross_val_score(lr, X_scaled, y, cv=5)

In [19]:
knn = KNeighborsClassifier()
knn_scores = cross_val_score(knn, X_scaled, y, cv=5)

In [20]:
dt = DecisionTreeClassifier()
dt_scores = cross_val_score(dt, X_scaled, y, cv=5)

In [21]:
rf = RandomForestClassifier()
rf_scores = cross_val_score(rf, X_scaled, y, cv=5)

In [22]:
svm = SVC()
svm_scores = cross_val_score(svm, X_scaled, y, cv=5)

In [24]:
results = pd.DataFrame({
    'Model': [
        'Logistic Regression',
        'KNN',
        'Decision Tree',
        'Random Forest',
        'SVM'
    ],
    'CV Score': [
        lr_scores.mean(),
        knn_scores.mean(),
        dt_scores.mean(),
        rf_scores.mean(),
        svm_scores.mean()
    ]})

results

,Model,CV Score
0,Logistic Regression,0.845854
1,KNN,0.833171
2,Decision Tree,1.000000
3,Random Forest,0.997073
4,SVM,0.921951


In [25]:
#Grid Search

from sklearn.model_selection import GridSearchCV

In [29]:
pg = {
    'n_estimators': [50,100,200],
    'max_depth': [3,5,10]
}

In [30]:
grid = GridSearchCV(
    estimator = RandomForestClassifier(random_state=42),
    param_grid = pg,
    cv=5, scoring='accuracy')

In [31]:
grid.fit(X, y)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
             param_grid={'max_depth': [3, 5, 10],
                         'n_estimators': [50, 100, 200]},
             scoring='accuracy')

In [32]:
print(grid.best_params_)

{'max_depth': 10, 'n_estimators': 50}


In [33]:
print(grid.best_score_)

0.9970731707317073


Model A: 92% accuracy
Model B: 89% accuracy

One train-test split can accidentally produce:

an easy test set
a difficult test set
a biased test set

Cross Validation solves this problem

Instead of evaluating the model once, we evaluate it multiple times.

K-Fold Cross Validation:

1025 patients
k = 5

Fold 1: 205 patient
Fold 2: 205 patient
Fold 3: 205 patient
Fold 4: 205 patient
Fold 5: 205 patient

Round 1:

Testing: Fold 1
Training: Fold 2,3,4,5

Round 2:
Testing: Fold 2
Training: Fold 1,3,4,5

Round 3... 4... 5...

Now every data point gets a chance to become testing data.

Instead of one score we get:

88%
90%
89%
91%
87%

Avg: 89%

Grid search is an exhaustive hyperparameter tuning technique used to find the optimal configuration for a machine learning model


max_depth = [2, 4, 6, 8]
n_estimators = [50, 100, 200]

(2,50), (2,100), (2,200), (4,50), (4,100), (4,200), ......
